### PR TITLE
Full propa test

### PR DESCRIPTION
**Pull request by tmontaigu**
### Check-list:

* [x] Tests for the changes have been added (for bug fixes / features)
5

### CODE REVIEW

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs [hunk 1]
```diff
@@ -3096,3 +3096,155 @@ where
...
+            "Invalid degree after propagation"
+        );
+    }
+
+    if block_carry_mod == block_msg_mod {
+        // This test is easier to write with this assumption
+        // which, conveniently is true for our radix type
+        //
+        // In this test, we are creating a ciphertext which is at full capacity
+        // which just enough room that allows sequential (non-parallel)
```
Comment (tmontaigu) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3157]: with just enough

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs [hunk 2]
```diff
@@ -3096,3 +3096,155 @@ where
...
+        );
+
+        // Take the initial value, but remove any bits below absober block
+        // as the bits below will have changed, but bits above will not.
+        let mut expected_built_by_hand =
+            initial_value & (u64::MAX << ((absorber_block_index + 1) as u32 * num_bits_in_msg));
+        // The first block generated a carry,
+        // but also results in a non zero block.
+        //
+        // The carry gets propagated until by other blocks
```
Comment (tmontaigu) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3242]: `get propagated by other blocks`

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs [hunk 3]
```diff
@@ -3096,3 +3096,159 @@ where
...
+        );
+        assert!(
+            ct.blocks
+                .iter()
+                .all(|b| b.degree.0 as u64 == block_msg_mod - 1),
+            "Invalid degree after propagation"
+        );
+    }
+
+    if block_carry_mod == block_msg_mod {
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3152]: block_carry_mod >= block_msg_mod ? I don't know if we have such parameters though

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs [hunk 4]
```diff
@@ -3096,3 +3096,159 @@ where
...
+        let msg = cks.encrypt(clear_max_value);
+        let mut ct = cks.encrypt(clear_max_value);
+        while sks.is_add_possible(&ct, &msg) {
+            sks.unchecked_add_assign(&mut ct, &msg);
+            expected_result = expected_result.wrapping_add(clear_max_value) % modulus;
+        }
+        assert!(ct
+            .blocks
+            .iter()
+            .all(|b| b.degree.0 as u64 == (block_total_mod - 1) - (block_msg_mod - 1)),);
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3171]: `- (block_carry_mod -1)` ?

**Review by IceTDrinker — COMMENTED** [review]
Will check the expected built by hand tomorrow some small comments

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs [hunk 5]
```diff
@@ -3096,3 +3096,159 @@ where
...
+        // but also results in a non zero block.
+        //
+        // The carry gets propagated by other blocks
+        // until it hits the absorber block, which takes the value of the carry
+        // (1) as its new value. Blocks that propagated the carry will have as new value
+        // 0 as for these block we did: ((block_msg_mod - 1  + 1) % block_msg_modulus) == 0
+        // and carry = ((block_msg_mod - 1  + 1) / block_msg_modulus) == 1
+        //
+        // Set the value of first block
+        expected_built_by_hand |= (2 * (block_msg_mod - 1)) % block_msg_mod;
```
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3249]: nevermind, my bad

**Comment by IceTDrinker** [comment]
looks good, just the minor comments for which I may be wrong

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs (hunk 4 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3171]: Yes I think `(block_carry_mod -1)` is the correct thing if we want to take into account  `block_carry_mod >= block_msg_mod`, but then In that case our full_propagate won't be a real full_propagate and so this test won't be correct

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs (hunk 3 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3152]: We have for shortints, but we don't use them in integers.

In integers params needs as many carry bit as message bits to have all operations
But if their are more carry bits than message bits, then we would need 2 full_propagation
to actually fully propagate.

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs (hunk 4 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3171]: Not sure about that, need to think about it 🤔

Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3171]: Ok so in the case where params are such that `block_carry_mod > block_msg_mod`, the sequential propagation is fine as long as is_op_possible do proper checks.

The parallel carry propagation would require to be adapted to support that case

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3171]: feel free to open follow up issues

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs (hunk 3 above)
Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3152]: ok so this is correct with our current implementation of full propagate for integers right ?

Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3152]: Ok Finally plugged my brain,

So using params where `block_carry_mod >= block_msg_mod`, the full_propagations (both sequential and parallel)
will correctly work, so as long as the block has enough space to receive the carry of the next block.
Which is what is_thing_possible where reworked to ensure (there is probably is_scalar_thing_possible left to update).

What confused me, is that the parallel carry propagation also works  by default in the even in the case where the whole block is completely full.

For example, in 2_2, the sequential propagation requires that there is enough room to receive worst case carry of preceding block, so max degree is 15 - 3 = 12. Using the parallel algorithm we can propagate carries where blocks
have a degree of 15 (ie degree == max_degree), however that's only true if `carry_modulus <= message_modulus`.

Parallel propagation could be improved to support the case where `carry_modulus  > message_modulus` and we sum until degree = (message_modulus * message_modulus)- 1.

Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3152]: I'm adding PARAM_MESAGE_2_CARRY_3 in the instanciated tests

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs]: ok !

**Review by IceTDrinker — APPROVED** [review]
Thanks ! Looks good !

**Comment by tmontaigu** [comment]
@slab-ci cpu_test

**Comment by tmontaigu** [comment]
@slab-ci cpu_integer_test

**Comment by tmontaigu** [comment]
Realized that the test was not meaningful as it did not check manually each block to verify its value (e.g no carry after propagation). Checking the result of the decryption is not a complete check, because decrypting propagtes carries, so an implemenation that did nothing would pass the tests

**Review by IceTDrinker — APPROVED** [review]
Looks good, and good catch on decryption doing the propagate automatically

**Comment by tmontaigu** [comment]
@slab-ci cpu_integer_test

**Comment by IceTDrinker** [comment]
@slab-ci cpu_integer_test

**Pull request merged**